spec/taskBehaviors.js

```diff
@@ -0,0 +1,267 @@
+        expect(runValues).toEqual([]);
+
+        // When running tasks, it will throw an exception and not complete the tasks
+        expect(function() {
+            jasmine.Clock.tick(1);
+        }).toThrow();
+        expect(runValues).toEqual([1]);
+
+        // The remaining tasks will run later
+        jasmine.Clock.tick(1);
+        expect(runValues).toEqual([1,2]);
+    });
+
+    it('Should stop recursive task processing after a fixed number of iterations', function() {
+        var runValues = [];
+        var func = function() {
+            runValues.push('x');
+            ko.tasks.schedule(func);
+        };
+
+        ko.tasks.schedule(func);
+        expect(runValues).toEqual([]);
+
+        expect(function() {
+            jasmine.Clock.tick(1);
+        }).toThrowContaining('Too much recursion');
+
+        // 5000 is the current limit in the code, but it could change if needed.
+        expect(runValues.length).toEqual(5000);
+    });
+
+    it('Should not stop non-recursive task processing', function() {
+        var runValues = [];
+        var func = function() {
+            runValues.push('x');
+        };
+
+        for (var i = 0; i < 10000; ++i) {
+            ko.tasks.schedule(func);
+        }
+        expect(runValues).toEqual([]);
+
+        jasmine.Clock.tick(1);
+        expect(runValues.length).toEqual(10000);
+    });
+
+    describe('Cancel', function() {
+        it('Should prevent task from running', function() {
+            var runCount = 0;
+            var handle = ko.tasks.schedule(function() {
+                runCount++;
+            });
+            ko.tasks.cancel(handle);
+
+            jasmine.Clock.tick(1);
+            expect(runCount).toEqual(0);
+        });
+
+        it('Should prevent only the canceled task', function() {
+            var runCount = 0;
+            var func = function() {
+                runCount++;
+            };
+            var handle1 = ko.tasks.schedule(func);
+            var handle2 = ko.tasks.schedule(func);
+            ko.tasks.cancel(handle2);
+
+            jasmine.Clock.tick(1);
+            expect(runCount).toEqual(1);
+        });
+
+        it('Should do nothing if task has already run', function() {
+            var runValues = [];
+            var func = function(value) {
+                runValues.push(value);
+            };
+            var handle1 = ko.tasks.schedule(func.bind(null, 1));
+            expect(runValues).toEqual([]);
+
+            jasmine.Clock.tick(1);
+            expect(runValues).toEqual([1]);
+
+            var handle2 = ko.tasks.schedule(func.bind(null, 2));
+
+            // Try to cancel the first task
+            ko.tasks.cancel(handle1);
+
+            // But nothing should happen; the second task will run in the next iteration
+            jasmine.Clock.tick(1);
+            expect(runValues).toEqual([1,2]);
+        });
+
+        it('Should work correctly after a task throws an exception', function() {
```

Comment:
I'm not confident that this behavior is what we'd ideally want. This test shows that it can cancel a task asynchronously, but *only* if one of the preceding but independently scheduled tasks throws an exception. If someone has behavior that relies on cancellation not occurring because it is attempted asynchronously, they will be surprised if sometimes cancellation does succeed because of some completely unrelated task throwing.

Overall I find the whole idea of task callbacks getting delayed by completely unrelated exceptions to be a bit concerning (e.g., as also specified on line 97 above), as it's unpredictable action-at-a-distance. Is this behaviour essential for any reason, or could we change it so that the rest of the queue continues to be processed synchronously even if one entry throws?